# origin: bk
# key: 41f59581p1swNaow4K1aBglV-q2jfQ (linux-2.5)
# description: packet forwarding DoS issue
# inclusion: projected 2.4.29 as backport
# revision date: 2005-02-16

# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/01/24 16:40:33-08:00 herbert@gondor.apana.org.au 
#   [IPV4/IPV6]: In ip_fragment(), reset ip_summed field on SKB sub-frags.
#   
#   If we forward a fragmented packet, we can have ip_summed
#   set to CHECKSUM_HW or similar.  This is fine for local
#   protocol processing, but once if we are forwarding this
#   packet we want to reset ip_summed to CHECKSUM_NONE.
#   
#   Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
#   Signed-off-by: David S. Miller <davem@davemloft.net>
# 
# net/ipv4/ip_output.c
#   2005/01/24 16:40:10-08:00 herbert@gondor.apana.org.au +1 -0
#   [IPV4/IPV6]: In ip_fragment(), reset ip_summed field on SKB sub-frags.
#   
#   If we forward a fragmented packet, we can have ip_summed
#   set to CHECKSUM_HW or similar.  This is fine for local
#   protocol processing, but once if we are forwarding this
#   packet we want to reset ip_summed to CHECKSUM_NONE.
#   
#   Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
#   Signed-off-by: David S. Miller <davem@davemloft.net>
# 
# net/ipv6/ip6_output.c
#   2005/01/24 16:40:10-08:00 herbert@gondor.apana.org.au +1 -0
#   [IPV4/IPV6]: In ip_fragment(), reset ip_summed field on SKB sub-frags.
#   
#   If we forward a fragmented packet, we can have ip_summed
#   set to CHECKSUM_HW or similar.  This is fine for local
#   protocol processing, but once if we are forwarding this
#   packet we want to reset ip_summed to CHECKSUM_NONE.
#   
#   Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
#   Signed-off-by: David S. Miller <davem@davemloft.net>
# 
diff -Nru a/net/ipv4/ip_output.c b/net/ipv4/ip_output.c
--- a/net/ipv4/ip_output.c	2005-02-16 15:58:30 -08:00
+++ b/net/ipv4/ip_output.c	2005-02-16 15:58:30 -08:00
@@ -504,6 +504,7 @@
 			/* Prepare header of the next frame,
 			 * before previous one went down. */
 			if (frag) {
+				frag->ip_summed = CHECKSUM_NONE;
 				frag->h.raw = frag->data;
 				frag->nh.raw = __skb_push(frag, hlen);
 				memcpy(frag->nh.raw, iph, hlen);
diff -Nru a/net/ipv6/ip6_output.c b/net/ipv6/ip6_output.c
--- a/net/ipv6/ip6_output.c	2005-02-16 15:58:30 -08:00
+++ b/net/ipv6/ip6_output.c	2005-02-16 15:58:30 -08:00
@@ -592,6 +592,7 @@
 			/* Prepare header of the next frame,
 			 * before previous one went down. */
 			if (frag) {
+				frag->ip_summed = CHECKSUM_NONE;
 				frag->h.raw = frag->data;
 				fh = (struct frag_hdr*)__skb_push(frag, sizeof(struct frag_hdr));
 				frag->nh.raw = __skb_push(frag, hlen);
